v2/rust/src/child_process_supervisor.rs

```diff
@@ -0,0 +1,93 @@
+use super::termination::TerminationFlag;
+use super::timeout::Timeout;
+use anyhow::{bail, Context, Result};
+use log::{error, warn};
+use std::collections::{HashMap, HashSet};
+use std::process::{Child, Command, ExitStatus};
+use std::thread::sleep;
+use std::time::Duration;
+use sysinfo::{Pid, PidExt, Process, ProcessExt, System, SystemExt};
+
+pub struct ChildProcessSupervisor<'a> {
+    pub command: Command,
+    pub timeout: u64,
+    pub termination_flag: &'a TerminationFlag,
+}
+
+impl ChildProcessSupervisor<'_> {
+    pub fn run(mut self) -> Result<ChildProcessOutcome> {
+        let mut child = self.command.spawn().context("Failed to spawn subprocess")?;
+        let timeout = Timeout::start(self.timeout);
+
+        loop {
+            if let Some(exit_status) = child
+                .try_wait()
+                .context(format!(
+                    "Failed to query exit status of process {}, killing",
+                    child.id()
+                ))
+                .map_err(|err| {
+                    kill_child_process_tree(&mut child);
+                    err
+                })?
+            {
+                return Ok(ChildProcessOutcome::Exited(exit_status));
+            }
+
+            if timeout.expired() {
+                error!("Process timed out");
+                kill_child_process_tree(&mut child);
+                return Ok(ChildProcessOutcome::TimedOut);
+            }
+
+            if self.termination_flag.should_terminate() {
+                warn!("Terminated");
+                kill_child_process_tree(&mut child);
+                bail!("Terminated")
+            }
+            sleep(Duration::from_millis(250))
+        }
+    }
+}
+
+pub enum ChildProcessOutcome {
+    Exited(ExitStatus),
+    TimedOut,
+}
+
+fn kill_child_process_tree(child: &mut Child) {
+    let mut system = System::new_all();
+    system.refresh_processes();
+    let _ = child.kill();
+    kill_all_children(&Pid::from_u32(child.id()), system.processes());
+}
+
+fn kill_all_children<'a>(top_pid: &'a Pid, processes: &'a HashMap<Pid, Process>) {
```

Comment:
Is this really the correct behaviour? From what I can see there are three things somebody can do:
* Send a SIGKILL signal such as this:
```
taskkill /pid 1230 /f
```
In such a case, `kill_all_children` would never be invoked, correct?
* Send a SIGKILL signal such as this:
```
taskkill /pid 1230 /f /t 
```
Again, taskkill is not invoked, but this time it doesn't matter, since all children are terminated by the invoker of `taskkill`.
* Send a SIGTERM, such as doing control-C on the command line. In this case, we need to invoking SIGTERM on each direct child only should be the correct behaviour. We then need to understand how rcc and robotframework handle such terminations.

Reply:
Windows simply doesn't have `SIGTERM`, I also confirmed this with Sergey. Otherwise, `SIGTERM` would be the way to go. `taskkill /t` does not work for our use cases, I think it only works with foreground applications, not sure though. It definitely cannot kill a Python process, I tried that.

Reply:
As discussed via Slack: I added documentation.